pkg/component/worker/ocibundle.go

```diff
@@ -56,46 +60,126 @@ func (a *OCIBundleReconciler) Init(_ context.Context) error {
+		a.Emit("can't read bundles directory")
+		return
+	}
+	a.EmitWithPayload("importing OCI bundles", files)
+	if len(files) == 0 {
+		return
+	}
 	for _, file := range files {
-		if err := a.unpackBundle(ctx, client, a.k0sVars.OCIBundleDir+"/"+file.Name()); err != nil {
-			a.EmitWithPayload("unpacking OCI bundle error", map[string]interface{}{"file": file.Name(), "error": err})
-			a.log.WithError(err).Errorf("can't unpack bundle %s", file.Name())
-			return fmt.Errorf("can't unpack bundle %s: %w", file.Name(), err)
+		fpath := filepath.Join(a.k0sVars.OCIBundleDir, file.Name())
+		a.log.Infof("Loading OCI bundle %s", fpath)
+		if err := a.loadOne(ctx, fpath); err != nil {
+			a.log.WithError(err).Errorf("Failed to load OCI bundle %s", fpath)
+			a.EmitWithPayload("failed to load OCI bundle", fpath)
+			continue
 		}
-		a.EmitWithPayload("unpacked OCI bundle", file.Name())
+		a.log.Infof("OCI bundle %s loaded", fpath)
 	}
 	a.Emit("finished importing OCI bundle")
+}
+
+// watch creates a fs watched on the oci bundle directory. This function calls load() anytime
+// a new file is created on the directory or a write operation took place . Events are debounced
+// with a timeout of 10 seconds. This function is blocking.
+func (a *OCIBundleReconciler) watch(ctx context.Context) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		a.log.WithError(err).Error("Failed to create watcher for OCI bundles")
+		return
+	}
+	defer watcher.Close()
+
+	if err := watcher.Add(a.k0sVars.OCIBundleDir); err != nil {
+		a.log.WithError(err).Error("Failed to watch for OCI bundles")
+		return
+	}
+
+	debouncer := debounce.Debouncer[fsnotify.Event]{
+		Input:   watcher.Events,
+		Timeout: 10 * time.Second,
+		Filter: func(item fsnotify.Event) bool {
+			switch item.Op {
+			case fsnotify.Create, fsnotify.Write:
+				return true
+			default:
+				return false
+			}
+		},
+		Callback: func(ev fsnotify.Event) {
+			a.log.Infof("Loading OCI bundle %s", ev.Name)
+			if err := a.loadOne(ctx, ev.Name); err != nil {
+				a.log.WithError(err).Errorf("Failed to load OCI bundle %s", ev.Name)
+				a.EmitWithPayload("failed to load OCI bundle", ev.Name)
+				return
+			}
+			a.log.Infof("OCI bundle %s loaded", ev.Name)
+		},
+	}
+
+	go func() {
+		for {
+			err, ok := <-watcher.Errors
+			if !ok {
+				return
+			}
+			a.log.WithError(err).Error("Error while watching oci bundle directory")
+		}
+	}()
+
+	a.log.Infof("Started to watch events on %s", a.k0sVars.OCIBundleDir)
+	if err := debouncer.Run(ctx); err != nil {
+		if errors.Is(err, context.Canceled) {
+			a.log.Info("OCI bundle watch bouncer ended")
+			return
+		}
+		a.log.WithError(err).Warn("OCI bundle watch bouncer exited with error")
+	}
+}
+
+// Starts initiate the OCI bundle loader. It does an initial load of the directory and
+// once it is done, it starts a watcher on its own goroutine.
+func (a *OCIBundleReconciler) Start(ctx context.Context) error {
+	a.loadAll(ctx)
+	go a.watch(ctx)
```

Comment:
I understand. My initial approach here was to run through the whole directory every time an event popped out (taking the necessary measures to avoid importing the same file twice). I have since changed to this new version as per @jnummelin suggestion.

Long story short, I will raise a new attempt at reviving the old approach + the changes you proposed here and ping you back.